_posts/2021-01-18-changelog-139.markdown

```diff
@@ -0,0 +1,113 @@
+---
+layout: post
+title: "IntelliJ Rust Changelog #139"
+date: 2021-01-18 17:00:00 +0300
+---
+
+
+## New Features
+
+* Cargo features and `cfg` support improvements:
+    <!-- https://github.com/intellij-rust/intellij-rust/pull/6497 -->
+    * Take into account attributes under [`cfg_attr`](https://doc.rust-lang.org/reference/conditional-compilation.html#the-cfg_attr-attribute) attribute during type inference, name resolution, and other analysis.
+      Now code insight works much better with code containing attributes like `#[cfg_attr(..., derive(...))]`, `#[cfg_attr(..., path="...")]`, etc.
+      TODO: gif
+
+    <!-- https://github.com/intellij-rust/intellij-rust/pull/6549 -->
+    * Fairly evaluate unknown `cfg` options to `false` if there is no custom [build script](https://doc.rust-lang.org/cargo/reference/build-scripts.html) (`build.rs`) in the package.
+      Previously, in order to avoid wrongly graying out the actual code, unknown `cfg` options and unknown Cargo features were considered _enabled_ and _disabled_ simultaneously.
+      Note, you can turn on the new logic for packages containing `build.rs` as well by enabling the experimental build script evaluation:
+      just enable `org.rust.cargo.evaluate.build.scripts` in `Help | Find Action | Experimental Features` dialog
+
+    <!-- https://github.com/intellij-rust/intellij-rust/pull/6507 -->
+    * Improve `cfg(test)` processing for external dependencies by providing a new heuristic.
+      Now `cfg(test)` is treated as _disabled_ if there is a `cfg(not(test))` condition in the project.
+      Previously, `test` option was considered _enabled_ and _disabled_ simultaneously, i.e. code under `cfg(test)` and `cfg(not(test)` was both enabled.
+      Such test attributes are quite complex to handle in the IDE: the tests should not be conditionally disabled since you might want to read and edit them at any time; but only mock objects should be grayed-out and excluded from code analysis.
```

Comment:
Still use `gray out` term here

Reply:
I have left both `conditionally disabled` and `grayed-out` in order to make the phrase clear enough for someone who potentially doesn't know what "conditionally disabled" means. Do you think it is very redundant?

Reply:
Maybe just add https://doc.rust-lang.org/reference/conditional-compilation.html for `conditionally` word and it solves this problem?
> Do you think it is very redundant?

I still think that `gray out` is about highlighting and it's not always the same as "code that the plugin considers disabled"

Reply:
Yes, I think the link might be helpful.

To be more precise, I'm afraid that the term "disabled" is not very common and understandable in this case. AFAIU the code may be _compiled_ or _included_, and `cfg` attribute "excludes" the code from the compilation. So "disabled code" might be not so clear. However, I don't think that "grayed-out code" is accurate enough too.